core/src/impl/Kokkos_TeamMDPolicy.hpp

```diff
@@ -0,0 +1,374 @@
+  // the second to the last nested level; otherwise, thread parallelism on the
+  // last nested level
+  static constexpr int left_par_rt =
+      (ThreadAndVector == TeamMDRangeThreadAndVector::Both) ? 1 : 0;
+
+  static constexpr int right_par_rt =
+      (ThreadAndVector == TeamMDRangeThreadAndVector::Both) ? Rank::rank - 2
+                                                            : Rank::rank - 1;
+
+  static constexpr int par_rt = is_direction_left ? left_par_rt : right_par_rt;
+
+  // Vector parallelism will always be on the last index
+  static constexpr int par_rv  = is_direction_left ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+
+// ThreadAndVectorNestLevel determines on which nested level parallelization
+// happens.
+//   - Rank is Kokkos::Rank<TotalNestLevel, Iter>
+//     - TotalNestLevel is the total number of loop nests
+//     - Iter is whether to go forward or backward through ranks (i.e. the
+//       iteration order for MDRangePolicy)
+//   - ThreadAndVector determines whether both vector and thread parallelism is
+//   in use
+template <typename Rank, typename ExecSpace,
+          TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel;
+
+#ifdef KOKKOS_ENABLE_SERIAL
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel<Rank, Serial, ThreadAndVector>
+    : HostBasedNestLevel<Rank, ThreadAndVector> {};
+#endif
+
+#ifdef KOKKOS_ENABLE_CUDA
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel<Rank, Cuda, ThreadAndVector>
+    : AcceleratorBasedNestLevel<Rank, ThreadAndVector> {};
+#endif
+
+#ifdef KOKKOS_ENABLE_HIP
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel<Rank, HIP, ThreadAndVector>
+    : AcceleratorBasedNestLevel<Rank, ThreadAndVector> {};
+#endif
+
+// TODO check par_rt and par_rv validity
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel<Rank, Experimental::OpenMPTarget,
+                                ThreadAndVector>
+    : HostBasedNestLevel<Rank, ThreadAndVector> {};
+#endif
+
+#ifdef KOKKOS_ENABLE_OPENMP
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel<Rank, OpenMP, ThreadAndVector>
+    : HostBasedNestLevel<Rank, ThreadAndVector> {};
+#endif
+
+#ifdef KOKKOS_ENABLE_THREADS
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel<Rank, Threads, ThreadAndVector>
+    : HostBasedNestLevel<Rank, ThreadAndVector> {};
+#endif
+
+// TODO check par_rt and par_rv validity
+#ifdef KOKKOS_ENABLE_SYCL
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel<Rank, Experimental::SYCL, ThreadAndVector>
+    : HostBasedNestLevel<Rank, ThreadAndVector> {};
+#endif
```

Comment:
Is there any reason this is different from` CUDA1/HIP` or have you just not tested?

Reply:
It just has not yet been tested for the `SYCL` backend. I changed it to be using the same config as `Cuda/HIP` in [ef6d137](https://github.com/kokkos/kokkos/pull/5238/commits/ef6d137d66aecc586bfe7b9ba673714a8978d40b).
Let me know if you think if this change is ok or if `SYCL` would benefit from a different configuration.

Reply:
If we are talking about multi-dimensional access, i.e., sycl::nd_range and sycl::nd_item, the order of access should be reversed to what we do for `CUDA`, compare
```
dim3 grid(int(m_league_size), 1, 1);
const dim3 block(int(m_vector_size), int(m_team_size), 1);
```
for `Cuda` with
```
sycl::nd_range<2>(
            sycl::range<2>(m_team_size, m_league_size * final_vector_size),
            sycl::range<2>(m_team_size, final_vector_size)),
```
for `SYCL`. Since the functionality added here is already nested inside a `TeamPolicy`, I would expect that doing the same as for `Cuda` (or `HIP` should be correct). It might be best to measure, though.

Reply:
It's fine to look at that later, though, if we do, we should create an issue to keep track of it.